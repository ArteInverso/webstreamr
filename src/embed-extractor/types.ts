import { StreamWithMeta } from '../types';

export interface EmbedExtractor {
  readonly id: string;

  readonly label: string;

  readonly extract: (embedUrl: string, language: string) => Promise<StreamWithMeta>;
}
