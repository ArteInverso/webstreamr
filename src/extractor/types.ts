import { Context, Meta, UrlResult } from '../types';

export interface Extractor {
  readonly id: string;

  readonly label: string;

  readonly ttl: number;

  readonly supports: (ctx: Context, url: URL) => boolean;

  readonly extract: (ctx: Context, url: URL, meta: Meta) => Promise<UrlResult[]>;
}
