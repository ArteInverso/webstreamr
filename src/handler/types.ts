import { Context, UrlResult } from '../types';

export interface Handler {
  readonly id: string;

  readonly label: string;

  readonly contentTypes: string[];

  readonly languages: string[];

  readonly handle: (ctx: Context, type: string, id: string) => Promise<(UrlResult | undefined)[]>;
}
