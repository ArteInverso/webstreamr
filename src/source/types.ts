import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Id } from '../utils';

export interface SourceResult {
  countryCode: CountryCode;
  title?: string;
  url: URL;
}

export interface Source {
  readonly id: string;

  readonly label: string;

  readonly contentTypes: ContentType[];

  readonly countryCodes: CountryCode[];

  handle(ctx: Context, type: ContentType, id: Id): Promise<(SourceResult[])>;
}
