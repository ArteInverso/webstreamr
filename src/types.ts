import { Manifest, ManifestConfig } from 'stremio-addon-sdk';

export interface Context {
  id: string;
  ip: string;
  config: Config;
}

export type ManifestWithConfig = Manifest & { config: ManifestConfig[] };

export type Config = Record<string, string>;

export interface Meta {
  bytes?: number;
  countryCode: string;
  height?: number;
}

export interface UrlResult {
  url: URL;
  isExternal?: boolean;
  label: string;
  sourceId: string;
  meta: Meta;
  requestHeaders?: Record<string, string>;
}
