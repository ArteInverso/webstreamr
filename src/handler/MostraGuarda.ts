import * as cheerio from 'cheerio';
import { Handler } from './types';
import { Fetcher, parseImdbId } from '../utils';
import { EmbedExtractorRegistry } from '../embed-extractor';
import { Context } from '../types';

export class MostraGuarda implements Handler {
  readonly id = 'mostraguarda';

  readonly label = 'MostraGuarda';

  readonly contentTypes = ['movie'];

  readonly languages = ['it'];

  private readonly fetcher: Fetcher;
  private readonly embedExtractors: EmbedExtractorRegistry;

  constructor(fetcher: Fetcher, embedExtractors: EmbedExtractorRegistry) {
    this.fetcher = fetcher;
    this.embedExtractors = embedExtractors;
  }

  readonly handle = async (ctx: Context, _type: string, id: string) => {
    if (!id.startsWith('tt')) {
      return [];
    }

    const html = await this.fetcher.text(ctx, new URL(`https://mostraguarda.stream/movie/${parseImdbId(id).id}`));

    const $ = cheerio.load(html);

    return Promise.all(
      $('[data-link!=""]')
        .map((_i, el) => new URL(($(el).attr('data-link') as string).replace(/^(https:)?\/\//, 'https://')))
        .toArray()
        .filter(embedUrl => !embedUrl.host.match(/mostraguarda/))
        .map(embedUrl => this.embedExtractors.handle(ctx, embedUrl, 'it')),
    );
  };
}
