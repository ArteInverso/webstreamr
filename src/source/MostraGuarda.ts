import { ContentType } from 'stremio-addon-sdk';
import * as cheerio from 'cheerio';
import { Source, SourceResult } from './types';
import { Fetcher, getImdbId, Id } from '../utils';
import { Context, CountryCode } from '../types';

export class MostraGuarda implements Source {
  readonly id = 'mostraguarda';

  readonly label = 'MostraGuarda';

  readonly contentTypes: ContentType[] = ['movie'];

  readonly countryCodes: CountryCode[] = [CountryCode.it];

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly handle = async (ctx: Context, _type: string, id: Id): Promise<SourceResult[]> => {
    const imdbId = await getImdbId(ctx, this.fetcher, id);

    const pageUrl = new URL(`https://mostraguarda.stream/movie/${imdbId.id}`);
    const html = await this.fetcher.text(ctx, pageUrl);

    const $ = cheerio.load(html);

    return Promise.all(
      $('[data-link!=""]')
        .map((_i, el) => new URL(($(el).attr('data-link') as string).replace(/^(https:)?\/\//, 'https://')))
        .toArray()
        .filter(url => !url.host.match(/mostraguarda/))
        .map(url => ({ countryCode: CountryCode.it, referer: pageUrl, url })),
    );
  };
}
