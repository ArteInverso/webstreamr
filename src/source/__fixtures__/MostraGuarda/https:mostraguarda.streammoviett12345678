not found tt12345678