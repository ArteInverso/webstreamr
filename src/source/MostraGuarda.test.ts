import winston from 'winston';
import { MostraGuarda } from './MostraGuarda';
import { Fetcher, ImdbId } from '../utils';
import { Context } from '../types';

jest.mock('../utils/Fetcher');

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const fetcher = new Fetcher(logger);
const ctx: Context = { id: 'id', ip: '127.0.0.1', config: { it: 'on' } };

describe('MostraGuarda', () => {
  let handler: MostraGuarda;

  beforeEach(() => {
    handler = new MostraGuarda(fetcher);
  });

  test('handles non-existent movies gracefully', async () => {
    const streams = await handler.handle(ctx, 'movie', new ImdbId('tt12345678', undefined, undefined));
    expect(streams).toHaveLength(0);
  });

  test('handle imdb the devil\'s bath', async () => {
    const streams = await handler.handle(ctx, 'movie', new ImdbId('tt29141112', undefined, undefined));
    expect(streams).toMatchSnapshot();
  });
});
