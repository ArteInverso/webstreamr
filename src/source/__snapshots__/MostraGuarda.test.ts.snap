// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MostraGuarda handle imdb the devil's bath 1`] = `
[
  {
    "countryCode": "it",
    "referer": "https://mostraguarda.stream/movie/tt29141112",
    "url": "https://supervideo.cc/e/6vri8jyhla3n",
  },
  {
    "countryCode": "it",
    "referer": "https://mostraguarda.stream/movie/tt29141112",
    "url": "https://dropload.io/embed-xsr90y2ltdyx.html",
  },
  {
    "countryCode": "it",
    "referer": "https://mostraguarda.stream/movie/tt29141112",
    "url": "https://mixdrop.ag/e/vk196d6xfzwwo1",
  },
  {
    "countryCode": "it",
    "referer": "https://mostraguarda.stream/movie/tt29141112",
    "url": "https://dood.to/e/37o04m33q3g5",
  },
]
`;
